packages/zip/lib/extract.js

```diff
@@ -11,15 +14,52 @@ const defaultOptions = {};
  * @param {Integer} options.defaultDirMode - Directory Mode (permissions), defaults to 0o755
  * @param {Integer} options.defaultFileMode - File Mode (permissions), defaults to 0o644
  * @param {Function} options.onEntry - if present, will be called with (entry, zipfile) for every entry in the zip
+ * @returns {Promise<{path: string}>}
  */
-module.exports = (zipToExtract, destination, options) => {
-    const opts = Object.assign({}, defaultOptions, options);
+module.exports = async (zipToExtract, destination, options) => {
+    if (options) {
+        // eslint-disable-next-line no-restricted-syntax
+        throw new Error('options are not supported anymore');
+    }
 
-    const extract = require('extract-zip');
+    if (!path.isAbsolute(destination)) {
+        // eslint-disable-next-line no-restricted-syntax
+        throw new Error('Target directory is expected to be absolute');
+    }
 
-    opts.dir = destination;
+    // Create the folder if it doesn't exist (same behaviour as extract-zip)
+    await fs.mkdir(destination, {recursive: true});
+    destination = await fs.realpath(destination);
 
-    return extract(zipToExtract, opts).then(() => {
-        return {path: destination};
+    const zip = new AdmZip(zipToExtract);
+
+    // Filter all __MACOSX folders and .DS_Store files
+    for (const entry of zip.getEntries()) {
+        if (entry.name === '.DS_Store' && !entry.isDirectory) {
+            zip.deleteFile(entry);
+        }
+    }
+
+    await new Promise((resolve, reject) => {
+        // adm-zip can call the callback multiple times, so we need to make sure we only resolve/reject once
+        let called = false;
+        zip.extractAllToAsync(destination, false, false, (err) => {
+            if (called) {
+                return;
+            }
+            called = true;
+            if (err) {
+                reject(err);
+            } else {
+                resolve();
+            }
+        });
     });
+
+    // Remove __MACOSX
+    const macosxPath = path.join(destination, '__MACOSX');
+    await fsExtra.remove(macosxPath);
```

Comment:
This is required because the `zip.getEntries()` method doesn't always return the `__MACOSX` folder.